ci/pipeline.yml

```diff
@@ -0,0 +1,305 @@
+  serial_groups: [mirror,proxy]
+  plan:
+  - in_parallel:
+    - get: registry-source
+      trigger: true
+      params: {depth: 1}
+      passed: [deploy-registry-staging]
+    - get: docker-registry-image
+      trigger: true
+      passed: [deploy-registry-staging]
+    - get: nginx-conf
+      trigger: true
+      passed: [deploy-proxy-staging]
+    - get: proxy-source
+      trigger: true
+      params: {depth: 1}
+      passed: [deploy-proxy-staging]
+    - get: ci-source
+      params: {depth: 1}
+  # Be sure we can pull an image through the mirror using the docker-image resource type
+  - task: test-pull-succeeds-docker-image
+    config:
+      platform: linux
+      image_resource:
+        type: docker-image
+        source:
+          repository: 18fgsa/concourse-task
+          registry_mirror: "https://((staging-mirror-hostname)).app.cloud.gov:443"
+      run:
+        path: /bin/sh
+        args:
+        - -ec
+        - |
+          date
+  # Be sure we can pull an image through the mirror using the registry-image resource type
+  - task: test-pull-succeeds-registry-image
+    config:
+      platform: linux
+      image_resource:
+        type: registry-image
+        source:
+          repository: 18fgsa/concourse-task
+          registry_mirror: 
+            host: "((staging-mirror-hostname)).app.cloud.gov:443"
+      run:
+        path: /bin/sh
+        args:
+        - -ec
+        - |
+          date
+    # on_failure:
+  - task: test-no-push
+    file: ci-source/ci/tasks/test-no-push.yml
+    params:
+      MIRROR_HOSTNAME: ((staging-mirror-hostname))
+  on_failure:
+    put: slack
+    params:
+      <<: *slack-params
+      text: |
+        :x: Integration tests for docker registry mirror failed in staging
+        <$ATC_EXTERNAL_URL/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|View build details>
+  on_success:
+    put: slack
+    params:
+      <<: *slack-params
+      text: |
+        :wave: Integration tests for docker registry mirror passed in staging. Production updates are ready to be deployed.
+        <$ATC_EXTERNAL_URL/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|View build details>
+
+- name: deploy-registry-production
+  serial_groups: [mirror]
+  plan:
+  - in_parallel: 
+    - get: registry-source
+      params: {depth: 1}
+      passed: [integration-test]
+    - get: docker-registry-image
+      passed: [integration-test]
+    - get: ci-source
+      params: {depth: 1}
+  - task: create-s3-bucket
+    file: ci-source/ci/tasks/create-s3-bucket.yml
+    params:
+      <<: *production-cf
+  - task: deploy-registry
+    file: ci-source/ci/tasks/deploy-registry.yml
+    params:
+      <<: *production-cf
+    on_failure:
+      try:
+        task: cancel-deployment
+        file: ci-source/ci/tasks/cancel-deployment.yml
+        params:
+          <<: *production-cf
+          CF_APP_NAME: docker-registry-mirror
```

Comment:
What happens if this rollback doesn't succeed?  I'm assuming we will have to dig in fairly quick, correct?  Will we know if the rollback failed in this case?

Reply:
cancel-deployment should re-deploy the last known good droplet. If the rollback failed, it would still be broken.

Reply:
That makes sense; we wouldn't really know the state until we checked though, right?  Then again we'd see the Slack notification, so we should be checking regardless; that'd let us know what state it's in and if we have time to investigate and fix or if a more immediate problem is on hand.

Reply:
That is the hope.